Responsive images: optimize the catalog images

vmarta opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

Ticket https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000BP0uXYAT/view

This PR takes advantage of the following items:
- Dynamic Image Service (DIS), which allows you to pass some parameters to easily transform your images
- `sizes` and `srcset` attributes to create the responsive images

With DIS, you can transform with these parameters, for example. For more details, see the [doc](https://documentation.b2c.commercecloud.salesforce.com/DOC1/index.jsp?topic=%2Fcom.demandware.dochelp%2Fcontent%2Fb2c_commerce%2Ftopics%2Fimage_management%2Fb2c_creating_image_transformation_urls.html).
- `q=60`: to compress the image with quality of 60%
- `sw=1000`: to resize the image to be 1000px wide

## `sizes` and `srcset`

If you're not familiar with what `sizes` and `srcset` are, here's a quick intro. `sizes` are like instruction we leave for the browser, while `srcset` is a pool of images with varying widths. The browser chooses the appropriate image width, based on `sizes` we gave them _and_ also the device pixel ratio (most mobile devices have at least 2x density, while 1x is commonly seen on desktop). For more details, see this [MDN page](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#resolution_switching_different_sizes).

An example:
```html
<img 
    src="https://example.com/image.jpg" 
    sizes="(min-width: 1200px) 800px, (min-width: 720px) 50vw, 100vw" 
    srcset="https://example.com/image.jpg?sw=800 800w, https://example.com/image.jpg?sw=1600 1600w" />
```

In the above code snippet, the `src` attribute is the fallback for very old browsers who don't support `sizes` or `srcset`. The `sizes` tells the browser _in order_: 
1. if the viewport is larger than 1200px, please use 800px image. 
2. Otherwise, if the viewport is at least 720px wide, select an image whose width is half the viewport's width. 
3. Or else, the image should fill up the whole viewport's width. 

Say, the viewport is currently at 1400px and you're using a mobile device with retina screen (2x density). The browser then would select an image with 800px * 2 = 1600px. It sees that there is a 1600px wide image in the `srcset`. If an exact resolution can't be found, it will choose the closest one.

## Challenges

I was initially not sure what the interface would be like. But inspired from this [Kent C. Dodds article](https://kentcdodds.com/blog/building-an-awesome-image-loading-experience#sizes-srcset-and-cloudinary), I decided to go with a utility function `getImageProps`. Since our page layout design aligns with the default Chakra breakpoints, `getImageProps` allows you to pass in `sizes` that have these breakpoints too.

### Value for `srcset`

Another challenge is what to use for the `srcset` value. Online articles typically gloss over this part. Understandably, it's because it's more like an art than science. It doesn't need to be 100% accurate, but should cover the spectrum of image widths quite well. 

One approach is to use the same set of image widths for every responsive image, while the other approach is more custom and manual. At the end, I decided to go with the latter and use this formula of 5 or so images: mobile 1x, 2x, 3x, and desktop 1x, 2x. Keep in mind too that the more images you provide, the less likely for a cache hit to happen.

### How to manually test

Testing the responsive images turns out to be tricky because of browser caching. Here's what I did to manually test the images:
1. Disable caching in the Network tab
2. Resize the viewport
3. Then reload the page ← if you don't reload, it seems that the browser prefers to use the highest-res image that it has downloaded recently
4. Inspect the image and hover your cursor on it to see what the "current source" is
5. Go to Step 2 and repeat, as you test for different devices

One thing to note is that the images we're working with are actually not great. They're quite pixelated for some reason. To see what I mean, compare these 2 images:
- https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw43751577/images/large/PG.10246370.JJBH3XX.PZ.jpg
- https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw43751577/images/large/PG.10246370.JJBH3XX.PZ.jpg?q=100

Even with or without quality of 100%, the background gradient of both images is not smooth. It looks as if the original images are already compressed too much.

### How much optimization

I found this article that outlines a [more complex optimization](https://jakearchibald.com/2021/serving-sharp-images-to-high-density-screens/) approach. It relies on varying the compression quality value for different device pixel ratio. For 2x images, you can actually use a lower quality value than with a 1x image. So that with 2x images, their file size doesn't balloon too quickly.

But I think the current implementation in this PR is good/simple enough for most developers. 

## Impact

Lighthouse score: I still need to check the lighthouse score. I don't think it'll be a huge change, though. Before the PR, the 'large' images are already at 800px wide, which is not that huge if you consider the high-density screens. Regardless, I predict that having responsive images is still a step forward in the positive direction.

## Remaining todos
- [x] Write tests for the util function
- [x] Look at more pages for other responsive images
- [x] Check lighthouse scores



# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [x] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`



# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [x] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [x] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

vmarta committed:
Draft of a util function to generate an image's srcset and sizes

vmarta committed:
Some refactoring

vmarta committed:
Tweak where getImageProps gets called

vmarta committed:
Set up responsive image for the image gallery

vmarta committed:
Some arguments are optional

Plus, adding some jsdoc.

vmarta committed:
Tweak image's fallback origin

vmarta committed:
Add some comments

vmarta committed:
Fix letter casing for React

vmarta committed:
Make `sizes.lg` more dynamic here

vmarta committed:
Responsive images for the product scrollers

vmarta committed:
Fallback src is now also transformed and with average width

vmarta committed:
Update jsdoc

vmarta commented on packages/pwa/app/components/product-scroller/index.jsx:
```diff
@@ -98,6 +98,14 @@ const ProductScroller = forwardRef(
+                                                    base: '70vw',
+                                                    md: '40vw',
+                                                    lg: '380px'
+                                                },
```
Instead of writing your own media queries for `sizes`, using an object like this is more consistent and convenient. It aligns you with the Chakra breakpoints.

vmarta commented on packages/pwa/app/components/product-tile/index.jsx:
```diff
@@ -70,7 +82,13 @@ const ProductTile = (props) => {
+                            src: image.disBaseLink,
+                            ...imageProps
+                        })}
+                    />
```
While ProductTile component knows how to get the image's alt and src, it doesn't and shouldn't know about the layout of the page it's used on. That's why `sizes` and `widths` props will be passed in through the `imageProps`.

vmarta commented on packages/pwa/app/pages/product-list/index.jsx:
```diff
@@ -390,6 +396,14 @@ const ProductList = (props) => {
+                                                  imageProps={{
+                                                      sizes: {
+                                                          base: 'calc(100vw / 2)',
+                                                          md: `calc((100vw - ${filtersContainerWidth}) / 3)`,
```
Your sizes don't have to be 100% accurate. Otherwise, the sizes would get complicated if you choose to measure the margins/paddings too. In this case, for `md` breakpoint, the width of a product tile would be roughly calculated as:
1. 100% of viewport width minus the filters container → you get the width of the main content
2. Then you divide it up into 3 columns

vmarta commented on packages/pwa/app/utils/responsive-image.js:
```diff
@@ -0,0 +1,89 @@
+    sizes.lg && s.push(`(min-width: 62em) ${sizes.lg}`)
+    sizes.md && s.push(`(min-width: 48em) ${sizes.md}`)
+    sizes.sm && s.push(`(min-width: 30em) ${sizes.sm}`)
+    sizes.base && s.push(sizes.base)
```
FYI the order matters for `sizes`, while it does not matter for `srcset`.

alexvuong replied:
Nit: is it possible to extract the breakpoint values to a constant object and reuse them here? I understand we follow Chakra default breakpoints, but it is possible that the customer may change the values for their own project, it is best to have a single place for those values.

vmarta commented on packages/pwa/app/utils/responsive-image.js:
```diff
@@ -0,0 +1,89 @@
+ *     })}
+ * />
+ */
+export const getImageProps = ({src, sizes, widths, transformations, ...otherProps}) => {
```
Here the `sizes` and `widths` are optional because:
- For more flexibility, let's not assume that every image is gonna be responsive images.
- Also, even if they're for responsive images, developers typically won't input them until later in development (when they know more about their responsive design).

vmarta commented on packages/pwa/app/components/image-gallery/index.jsx:
```diff
@@ -101,7 +102,19 @@ const ImageGallery = ({imageGroups = [], selectedVariationAttributes = {}, size}
+                            {...getImageProps({
+                                alt: heroImage.alt,
+                                src: heroImage.disBaseLink,
+                                sizes: {
+                                    base: '100vw',
+                                    lg: styles.heroImage.maxWidth[3]
+                                },
```
While ProductTile component does not know about the `sizes` and `widths`, the ImageGallery knows the values for them. The difference here is that ImageGallery is more like a bigger component that knows about its layout on the page.

bendvc commented on packages/pwa/app/utils/responsive-image.js:
```diff
@@ -0,0 +1,89 @@
+ * />
+ */
+export const getImageProps = ({src, sizes, widths, transformations, ...otherProps}) => {
+    const url = transformImageUrl(src, transformations)
```
I'm not sure how I feel about this average logic. Is the additional complexity worth the end result? I'm not sure, my first assumption would be that the default might have no transformations at all. In terms of defining your jsdocs, saying "src will be used as the fallback url if your browser doesn't support image sets" is a lot easier than "the default image will be the average width of the values passed in the widths array if you browser doesn't support image sets". 

Also, this calculation would probably have a negative impact on caching if I was to hazard a guess.

bendvc commented on packages/pwa/app/pages/product-list/index.jsx:
```diff
@@ -390,6 +396,14 @@ const ProductList = (props) => {
+                                                          md: `calc((100vw - ${filtersContainerWidth}) / 3)`,
+                                                          '2xl': '387px'
+                                                      },
+                                                      widths: [400, 800, 1200]
```
Is this a true statement base on the configuration that we have above? 

If the screen is determined to be "2xl" based on Chakra responsive layout, we want to use a image with a width of 384px, that will result in the image set image of w=400 based on the browsers internal matching algorithm. If the largest screen size is using the smallest image in the set, why do we have widths of 800 and 1200. Is it because of the pixel density that those smaller screen sizes might have?

vmarta replied:
@bendvc yes, that's right. The 800 and 1200 are for 2x and 3x pixel density screens.

Heads up: I've just updated the widths values. I'll explain how I got to those values in another comment.

vmarta committed:
`src` is no longer transformed

vmarta committed:
Revisit the widths calculation

My rule of thumb is mobile 1x, 2x, 3x and desktop 1x, 2x.

Mobile is observed with Chrome's Mobile Large breakpoint (425px), while Desktop is around 1600px (where Chakra 2xl breakpoint is).

vmarta commented:
@bendvc I've just updated the `widths` values. My rule of thumb is still mobile 1x, 2x, 3x, and desktop 1x, 2x. I wasn't being consistent in how I measured mobile and desktop, though. 

So now I define _mobile_ as Chrome's Mobile Large breakpoint (425px), while _desktop_ is around 1600px (where Chakra's 2xl breakpoint is at).

For the product list, the calculation would be:
- mobile 1x, 2x, 3x: 189, 378, 567
- desktop 1x, 2x: 387, 774

You see that 378 and 387 are very close to each other, so we'll take the bigger number. The end result is then `[189, 567, 387, 774]`.

vmarta committed:
Fallback if DIS url for an image is undefined

alexvuong commented on packages/pwa/app/components/image-gallery/index.jsx:
```diff
@@ -101,7 +102,19 @@ const ImageGallery = ({imageGroups = [], selectedVariationAttributes = {}, size}
                     <AspectRatio {...styles.heroImage} ratio={1}>
-                        <Img alt={heroImage.alt} src={heroImage.disBaseLink} />
+                        <Img
+                            {...getImageProps({
```
Nit: I'd prefer we avoid using a spread operator for a function like this. It makes the component props a bit too long. Can we create a variable `heroImageProps` = `getImageProps()` above and use it here instead?

bendvc commented on packages/pwa/app/components/product-tile/index.jsx:
```diff
@@ -70,7 +82,13 @@ const ProductTile = (props) => {
+                    <Img
+                        {...getImageProps({
+                            alt: image.alt,
+                            src: image.disBaseLink || image.link,
```
Will this logic result in the standard link having all the dis transformations on it?

vmarta replied:
@bendvc with my latest changes, the `src` value is no longer transformed, regards if it's a `disBaseLink` or `link`.

alexvuong commented on packages/pwa/app/components/product-scroller/index.jsx:
```diff
@@ -98,6 +98,15 @@ const ProductScroller = forwardRef(
+                                                    md: '40vw',
+                                                    lg: '30vw',
+                                                    '2xl': '489px'
+                                                },
```
Dumb question: why do we use px for `2xl` while the rest is `vw`?

vmarta replied:
@alexvuong that's because for `2xl`, the images no longer growing bigger and they're fixed size. So I was able to use a px unit for them.

There's a pattern to our page layout design in general: at '2xl', the main content no longer stretches because there's a max-width assigned to the page. As a result, the images inside also do not grow anymore.

alexvuong commented on packages/pwa/app/components/product-scroller/index.jsx:
```diff
@@ -98,6 +98,15 @@ const ProductScroller = forwardRef(
+                                                    lg: '30vw',
+                                                    '2xl': '489px'
+                                                },
+                                                widths: [275, 550, 825, 489, 978]
```
I am concerning of passing these numbers as `widths` directly as prop here. They look pretty "magical". 
Can we put them into a style prop and reuse them here? or Is there a way to make them more sense to pass as widhtProps (perhaps making they array into a constant with comment explaining the values)?

vmarta replied:
@alexvuong yeah, I can see how `widths` can be magical 😅 Let me write down what Ben suggested to me earlier during a huddle. It would try to make the interface simpler, so the widths would be automatically calculated for you.

alexvuong commented on packages/pwa/app/pages/product-list/index.jsx:
```diff
@@ -183,6 +183,8 @@ const ProductList = (props) => {
     /**************** Filters ****************/
     const [searchParams, {stringify: stringifySearchParams}] = useSearchParams()
     const [filtersLoading, setFiltersLoading] = useState(false)
+    const filtersContainerWidth = '280px'
```
Want to move this outside of the component as it is a constant?

vmarta commented:
Ben suggested to me earlier to try having a simpler interface that other developers can easily use. It would rely on viewport width, which allows you to define your image in terms of columns (1-column layout, 2-cols, 3-cols, etc.) I'm thinking something like this:

```jsx
const image = {link: '', disBaseLink: ''}

            {/* This image would be 50vw across all breakpoints */}
            <DISImg {...image} viewportWidth="50" />
            {/* Mimicking Chakra's responsive styles (in both array and object formats) */}
            <DISImg {...image} viewportWidth={[100, 100, 100, 50]} />
            <DISImg {...image} viewportWidth={{base: 100, lg: 50}} />
```

And the `viewportWidth` would get mapped to an image's `sizes` and `srcSet` attributes automatically. Although this is a simpler interface, the tradeoff is that this method is less accurate than before. It is a rough approximation. If you want to be more granular, `getImageProps` util function is still available, and it will be used to implement the simpler interface.

About the name `viewportWidth`:
- First of all, it's singular and not plural. This is following Chakra's style props.
- Chakra already has the built-in `width` prop, whose integer value is in pixel unit. So I thought that since this is dealing with vw unit, we can call it `viewportWidth`.

cc @bendvc @alexvuong

vmarta committed:
Brainstorm the simpler interface

vmarta commented:
After chatting with Ben last time, here's another iteration, where:
- The interface is more flexible.. able to accept non-DIS images
- And yes, it comes with some complexity too

## Now a component

Ben wanted something that's more React-like. Because currently our codebase uses both of Chakra's `<Image>` and `<Img>` components, I propose to have the same named components in our PWA. They will output to the corresponding Chakra's components.

Why not name them as `<ResponsiveImage>` or `<ImageSet>`? Because if I were a component, I should not know whether the layout of the page I'll be used on is responsive or not. And thus, I can't assume that I will render a responsive image too. I'll need to be able to render either a regular or responsive image.

## A simpler interface

In this simpler interface, users would only need to pass in `vwSizes` prop, which then would be automatically mapped to the image's `sizes` and `srcset` attributes.

But if they do need more advanced interface, they can still pass in `sizes` and `srcSet` props:
- `sizes` accepting either an object or array ← this mimics Chakra's responsive style props
- `srcSet` accepting an array

If they need to have complete control, they can simply pass in _literal strings_ for the `sizes` and `srcSet`. In this case, the props will be passed through to the Chakra's components.

## Examples
Here are some examples of the new interface I'm proposing. Let me know what you think of it @bendvc @alexvuong and we can huddle too.

```jsx
    // Update (Nov 15): `src` now contains the url transformations in it (e.g. `q=60`)
    const DISImage = (
        <Image
            // With the special syntax `[{}]`, this `src` will be mapped to
            // the fallback src attribute (for old browser that doesn't support srcSet)
            // of "https://edge.disstg.commercecloud.salesforce.com/...jpg" (without ?sw=)
            src="https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw1e4fcb17/images/large/PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}&q=60]"
            // `vwSizes` will be automatically mapped to sizes and srcSet attributes
            // and thus less work for the users when implementing responsive images.

            // Why I chose this name?
            // - 'vw', because the values are in vw unit (not px)
            // - 'Sizes' and not width/srcset, because the values correspond to responsive breakpoints
            vwSizes={[100, 100, 100, 50]}
            // OR vwSizes={{base: 100, lg: 50}}
        />
    )

    // Flexible enough to support other kinds of image urls
    const nonDISImage = (
        <Image
            // This is a static bundled image
            src="https://example.com/image[_{width}].jpg" // e.g. image_720.jpg
            
            // While an image service allows us to use any sized images, we can't do that with static bundled images.
            // Thus we can't accept `vwSizes` here but we can let the users manually specify the sizes and srcSet.
            sizes={['100vw', '100vw', '50vw', '350px']} // Remember: sizes relate to Chakra breakpoints
            srcSet={[300, 720, 1000, 1500]} // srcSet values will be interpreted in px unit (which is what the standard specifies)
        />
    )

    // Mimicking Chakra's style prop, you can pass in either object or array for `sizes`
    const sizesWithObjectOrArray = (
        <Image
            src="https://example.com/image[_{width}].jpg"
            sizes={{
                base: 'calc(100vw / 2)',
                md: `calc((100vw - 280px) / 3)`,
                '2xl': '387px'
            }}
            srcSet={[189, 567, 387, 774]}
        />
    )

    // Can pass in literal, handcrafted values
    const withLiteralValues = (
        <Image
            src="https://example.com/image.jpg"
            sizes="(min-width: 1200px) 800px, (min-width: 720px) 50vw, 100vw"
            srcSet="https://example.com/image_800.jpg 800w, https://example.com/image_1600.jpg 1600w"
        />
    )

    // No sizes or srcSet attributes? No problem.
    const nonResponsiveImage = <Image src="https://example.com/image.jpg" />
```

vmarta committed:
Convert to components

vmarta commented:
For the **lighthouse tests**, it's weird that Chrome seems to not able to generate the report. It just hanged for some reason for me. Anyways, I ran the tests via the command line `npm run test:lighthouse` instead.

The result is that there isn't much improvement in the scores. The page where there's a visible improvement is the image-heavy Product List page:
- Before: https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1636405786444-47167.report.html
- After using responsive images: https://storage.googleapis.com/lighthouse-infrastructure.appspot.com/reports/1636404368732-12358.report.html

Note the bump in Performance score (from 48 to 51). Specifically, when you compare "Opportunity > Properly size images", you'll see there's an improvement there.

I think a greater performance impact will be felt when implementing lazy-load images, especially for an image-heavy page like PLP.

Keep in mind too that the reports complained about Javascript. Please ignore that for now, because the Javascript code is not minified when viewing localhost.

alexvuong commented:
@vmarta I quick thought on the new iteration proposal. 

- Instead of having transformation and vsSizes as direct prop, we can group them inside of a props say `DISImgeProps ` and put anything required to generate sizes and srcset for DISImage, pass them to chackra image. For more safer guard, maybe create a flag to indicate whether it is a DISimage or not => This will make it easier to read the prop for the image component
- For any nonDIS image, they can just use three props (src, sizes, srcSet) or the minimum src

vmarta commented:
> * Instead of having transformation and vsSizes as direct prop, we can group them inside of a props say `DISImgeProps ` and put anything required to generate sizes and srcset for DISImage, pass them to chackra image. For more safer guard, maybe create a flag to indicate whether it is a DISimage or not => This will make it easier to read the prop for the image component

@alexvuong here's a comparison of before and after. I'm not sure if it makes much difference in terms of readability:

```jsx
    const a = (
        <Image
            src="https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw1e4fcb17/images/large/PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}]"
            transformations={{
                q: 60
            }}
            vwSizes={[100, 100, 100, 50]}
        />
    )

    const b = (
        <Image
            src="https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw1e4fcb17/images/large/PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}]"
            DISImage={{transformations: {q: 60}, vwSizes: [100, 100, 100, 50]}}
        />
    )
```

Grouping the props reduces the number of total props from 3 to 2. And by grouping `transformations` and `vwSizes`, I think it becomes more obscured/hidden. So my preference would be having them as separate props.

alexvuong commented:
> Grouping the props reduces the number of total props from 3 to 2. And by grouping `transformations` and `vwSizes`, I think it becomes more obscured/hidden. So my preference would be having them as separate props.

The reason I suggest grouping it into a `DISImageProps` is that if a user declares it, it is to indicate a DIS Image is being used and anything inside the object is for DISImage. I don't need to know what is technically inside the object to know this image is using a dis image. I know we can tell based on the url, but url can be changed anytime. 
If we have them as separate props, the user won't know the `vwSize` and `transformations` are for DISImage usage only until they go to the image and read about them. 

Image if I pass in a non-dis image, but I have transformations and vwSizes. It might not work of course, but on the readability aspect, it does not tell me anything until I have to go inside the component, read what `transformations` and `vwSizes` are for since they are the new "props" we introduce to the component. 

```
<Image
    src="https://example.com/image.jpg""
    transformations={{
        q: 60
    }}
    vwSizes={[100, 100, 100, 50]}
/>
```

As for clarifying the obscurity, we can try to use propType to create the shape of the `DISImageProp` Object

```
Image.propTypes = {
   DISIMageProps: PropTypes.shape({
        vwSize: PropTypes.array.isRequired,
        transformations: PropTypes.object
    }).isRequired,
}
```

vmarta commented:
@alexvuong actually, we don't need to pass in `transformations` prop. We can simplify the interface like this:

```jsx
<Image
            src="https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw1e4fcb17/images/large/PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}&q=60]"
            vwSizes={[100, 100, 100, 50]}
        />
```

Note that the `src` value now contains the url transformations in it: `.../PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}&q=60]`.

So now we only have 1 new prop called `vwSizes`. And again, it will be automatically mapped to the corresponding `sizes` and `srcset` attributes.
- Chakra's <Image> component supports both `sizes` and `srcset` as _separate_ props. Thus, I'd argue that `vwSizes` be a separate prop too.
- How would users know what `vwSizes` is for? Yes, they would need to look it up in our jsdoc or image component. It's very similar to users looking up the props for any Chakra components.

When you get back from your vacation, can we huddle about this? I'd love to hear more about your concern.

vmarta committed:
No more transformations

alexvuong commented:
> @alexvuong actually, we don't need to pass in transformations prop. We can simplify the interface like this:
> 
> <Image
>             src="https://edge.disstg.commercecloud.salesforce.com/dw/image/v2/ZZRF_001/on/demandware.static/-/Sites-apparel-m-catalog/default/dw1e4fcb17/images/large/PG.10212867.JJ3XYXX.PZ.jpg[?sw={width}&q=60]"
>             vwSizes={[100, 100, 100, 50]}
>         />

It looks simpler this way, I find it okay to do so taking into account we document it well in the component, and clarify which props get overridden when this prop is being used.

vmarta committed:
Coding the image component interface

vmarta committed:
First draft of the new interface for responsive image

vmarta committed:
Implement `getSrc`

vmarta committed:
No longer hardcoding the breakpoints labels

vmarta committed:
Revert "No longer hardcoding the breakpoints labels"

This reverts commit dd6942c2ff7dad963dbffc9f58a7857da72c89af.

vmarta committed:
Second draft of the new interface

vmarta committed:
srcSet for higher pixel density

vmarta committed:
Move functions into util

And also refactored some of other components and pages.

vmarta committed:
Add some comments

bendvc commented on packages/pwa/app/components/image/index.jsx:
```diff
@@ -0,0 +1,196 @@
+    srcSet: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.number), PropTypes.string])
+}
+
+export const Image = ({src, vwSizes, sizes, srcSet, ...otherProps}) => {
+    return <ChakraImage {...getImageProps({src, vwSizes, sizes, srcSet, ...otherProps})} />
+}
+Image.propTypes = propTypes
```
A few comments on this block: 

1. It looks like we are passing `otherProps` to the `getImageProps` utility function where it's unused, is there a specific reason we are passing it though? 
2. We are also passing a prop called `srcSet` which is modified and returned with the same name. Is there a better name for this prop so it's easily differentiated from the value (aptly named srcSet because it is a html srcSet) that comes back?
3. I renamed `getImageProps` to `getResponsiveImageAttributes` to be a little more descriptive. But thats just a suggestion.

```suggestion
export const Image = ({src, vwSizes, sizes, srcSet, ...rest}) => {
    return <ChakraImage {...rest} {...getResponsiveImageAttributes({src, vwSizes, sizes, srcSet})} />
}
Image.propTypes = propTypes
```

vmarta replied:
Ok, I've committed your suggestions for 1 and 3.

vmarta replied:
For 2: I was thinking too of a different prop name, but decided that `srcSet` and `sizes` are good names already and their values can be represented in different formats, similar to [Chakra's responsive styles](https://chakra-ui.com/docs/features/responsive-styles#the-array-syntax). 

I see our Image component as an extension of Chakra's Image component. And in Chakra, for example, style props like `width` can be represented in any of these values:
- literal string → it will pass through without modification
- array
- object

vmarta replied:
@bendvc On second thought, I think you got a point there (for number 2). How about renaming the props to these?

- `sizesForBreakpoints` → mapped to `sizes` - I think now it's clearer that the sizes are actually relative to the breakpoints
- `srcSetWidths` → mapped to `srcSet`

vmarta replied:
And about the existing `vwSizes`: I now think it's not necessary and we can remove it to make things clearer.

Instead of `vwSizes={[100, 100, 50]}`:
```jsx
sizesForBreakpoints={['100vw', '100vw', '50vw']}
// If `srcSet` or `srcSetWidths` are not given, we would automatically calculate it for them
```

vmarta committed:
Add tests (and fix a bug too)

vmarta committed:
Fix a test

vmarta committed:
Address Ben's feedback

vmarta committed:
Some small refactoring

vmarta committed:
Simplify image props with `vwSizes`

vmarta committed:
Update jsdoc

vmarta commented:
Ok @bendvc , I've addressed your existing feedback. Let me know what other feedback you have, thanks.

vmarta committed:
Merge branch 'develop' into responsive-images

vmarta committed:
Update CHANGELOG.md

vmarta committed:
Remove temporary code

vmarta committed:
Fix linting error

vmarta committed:
Update jsdoc

vmarta committed:
Update jsdoc, prop types

bendvc commented:
> Ok @bendvc , I've addressed your existing feedback. Let me know what other feedback you have, thanks.

Hey @vmarta, I finally was able to check out your code run it and do a deeper dive in to how things work. Here are a few comments that we can open up discussion about.

1. We seem to be doing a lot of name overloading, whether it's the component name itself or the props that component implements. This was really confusing to me, and I think it might make the components API difficult to use/understand. I think I mentioned this before and you recently came up with some alternatives. Below is a suggestion for a simple API

```
// Basic usage with incomplete widths array (will use default units, and missing breakpoints will be follow 
// how chakra does its responsive value types.
<ImageSet src="https://www.imagehorder.com/house.jpg?w={width}" widths=[100, 200] />

// Basic usage with complete list of widths for all breakpoints.
<ImageSet src="https://www.imagehorder.com/house.jpg?w={width}" widths=[100, 100, 100, 100, 200] />

// More advanced usage using object map for breakpoints and width values (incomplete object)
<ImageSet src="https://www.imagehorder.com/house.jpg?w={width}" widths={{base: 100, xl: 500}} />

// Same as above but with defined units in values.
<ImageSet src="https://www.imagehorder.com/house.jpg?w={width}" widths={{base: '50vw', xl: '100vw'}} />

// By default we'll use the Img component, if you want to use the Image component by passing it via the `as`
// prop. If you want to control the props for that component maybe you can access them via `imgProps`. This 
// example also uses a different component name that might be better.
<DynamicImage src="https://www.imagehorder.com/house.jpg?w={width}" widths=[100, 200] as={Image} imgProps={}/>
```

_NOTE: Doing something like above has a limitation in that you can't have completely different images for different breakpoints, so maybe ImageSet is a bad name (Maybe DynamicImage is a better name, kinda like DIS but not??). But the advantage of not using Image is that we won't have all those property name conflicts forcing use to come up with long property names that are hard to name. It also breaks us away from having to support all those props for Img/Image implicitly meaning we won't have the one util function getResponsiveImageAttributes as a gate keeper for all props for the image component because it's no longer our concern. _

_NOTE: By adding the `as` property it will allow to not export 2 component in a single component definition which doesn't really follow the larger overall pattern we are trying to achieve. _

2. Chakra utilizes a utility function called `mapResponsize` to achieve some of the things you are doing with mapping props. I'd suggest that you take a look at how they use it and see if thats something we can do, [here](https://github.com/chakra-ui/chakra-ui/blob/main/packages/layout/src/simple-grid.tsx) is a place in their code were they use it, it looks like something we are doing. (Less coding for use and less testing).

3. Similar to above we might want to reuse some of Chakras types if possible, e.g. in my example `widths` would be a ResponsiveValue type.

vmarta commented:
Before we meet to discuss, let me write down my thoughts.

`<ImageSet>`:
- Having it implies that we know ahead of time that the image is going to be responsive image. For example with ProductTile component, it shouldn't know whether the page it's on is responsive or not. Thus, we need to rely on the parent component to let us know whether it's a regular or responsive image.
- Yeah, what's a better name for this component? If it's DynamicImage, what happens to supporting static images?

`widths`:
- I think the name of this props is not specific enough, since there's the concept of widths in both sizes and srcset. 
- We would also need to have an additional prop (srcset) for supporting static images.

default unit: What is the default unit?

`src` url: we'll need a way to indicate which portion of it is optional

Chakra utils: 
- It's not meant for public use.
- What's the benefit of using Chakra types?

vmarta commented:
After our discussion, we've decided to reduce the scope and functionalities to focus on supporting dynamic images, where:

- You can simply pass in widths relative to the breakpoints, and these widths will then be mapped to the corresponding `sizes` and `srcSet` for the Chakra's `Img` component.
- The widths can be either number (by default, it means px value) or string (only vw and px units are accepted, as they're what can be converted to px on the server side)
- The resulting `srcSet` would include 2x versions for the higher pixel density devices

vmarta committed:
Reduce functionalities to focus on just the dynamic image

vmarta committed:
Fix a few bugs and write tests

vmarta committed:
Rename file

vmarta committed:
Implement `as` prop and add jsdoc

vmarta committed:
Remove unnecessary width param

vmarta committed:
Update comments

vmarta committed:
Merge branch 'develop' into responsive-images

vmarta committed:
Update comments

vmarta committed:
Merge branch 'develop' into responsive-images

vmarta commented:
@bendvc let me know when you've had a chance to review the PR, thanks.

bendvc commented on packages/pwa/app/utils/responsive-image.js:
```diff
@@ -0,0 +1,179 @@
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import theme from '@chakra-ui/theme'
```
This theme value is Chakra's default theme, it's not necessarily the theme that the project is using. You can see that we are extending the theme in `app/theme/index.js`. Although we aren't extending the `breakpoints` in our theme, and using the defaults, there is a pretty decent change that someone will want to change those values through theme extension. 

This means that our component won't react properly to those changes. An idea you can look into is to use the breakpoints from the actual theme as an argument. You'd probably have to get the theme via the `useTheme` hook and use it in your component and pass the breakpoint through `getResponsiveImageAttributes`.

vmarta replied:
Good catch! I've now used `useTheme` to get the app theme's breakpoints.

bendvc commented on packages/pwa/app/components/dynamic-image/index.jsx:
```diff
@@ -0,0 +1,49 @@
+ * <DynamicImage src="http://example.com/image.jpg[?sw={width}&q=60]" widths={['50vw', '100vw', '500px']} />
+ */
+const DynamicImage = ({src, widths, imageProps, as, ...rest}) => {
+    const Component = as ? as : Img
```
So I was digging around and I noticed that the utility function used to get the responsive image properties was getting called a lot (on the pdp), as it turns out there is probably something wrong with the page itself because the hero image in the gallery was re-rendering a lot casing this component to mount and unmount a lot.

I initially thought that there was something up with this component and thought that using the useMemo hook to only compute these values once would fix it like this:

```
    const responsiveProps = useMemo(() => getResponsiveImageAttributes({src, widths}), [
        src,
        widths
    ])
```

But because it's actually being mounted and unmounted from it's parent there is nothing you can do. There is no real action here. But I just wanted to show you my though pattern and its usually a good idea to use the useMemo hook for operations like this.

bendvc replied:
Another crazy thought...

Is it possible that because we don't re-calculate the responsive attributes on browser size change, that we get into a weird state where we have pixelated images for a specific window width, or conversely an overly large image for a small window?  

I don't expect you to fix that right now, but if it's the case then maybe we can have a known issue with a follow up.

vmarta replied:
Yes, `useMemo` is still useful for a complex operation like `getResponsiveImageAttributes`. I've now used it.

vmarta replied:
Regarding `we don't re-calculate the responsive attributes on browser size change`, as we've discussed last Friday, this is not a worry because we calculate based on the theme's breakpoints (unlike window's width, they're available on the server-side).

vmarta committed:
Get the correct breakpoints from current theme

vmarta committed:
Easier breakpoint logic to reason with

vmarta committed:
Memoize the result of getResponsiveImageAttributes

vmarta committed:
Update jsdoc

vmarta committed:
Move code block closer to where it's being used

vmarta committed:
Test passing in theme breakpoints

vmarta commented:
@bendvc this PR is ready for another review, thanks.

bendvc committed:
Merge branch 'develop' into responsive-images

vmarta committed:
Merge branch 'develop' into responsive-images

vmarta committed:
Merge branch 'develop' into responsive-images

sejal-salesforce commented:
@rsexton404 I know we haven't yet documented components, but this one is a great candidate to document now to use DIS with DynamicImage component. @vmarta has added very useful info that can go to docs from here.
